### PR TITLE
Display preview image from file data instead of url, handle failover

### PR DESCRIPTION
**Pull request by jbalboni**
In progress changes to use in memory image data for previews, and handle errors if we request an image from the backend that is still processing and fails.

screen shot 2018-02-19 at 2 01 38 pm

### CODE REVIEW

**Review by cehsu — APPROVED** [review]

File: src/js/vic-v2/components/PhotoField.jsx [hunk 1]
```diff
@@ -210,6 +216,17 @@ export default class PhotoField extends React.Component {
...
   }
 
+  componentWillReceiveProps(nextProps) {
+    const nextFormData = nextProps.formData || {};
+    const prevFormData = this.props.formData || {};
+    if (nextFormData.file instanceof Blob && nextFormData.file !== prevFormData.file) {
+      if (this.state.previewSrc) {
+        window.URL.revokeObjectURL(this.state.previewSrc);
+      }
+      this.setState({ previewSrc: window.URL.createObjectURL(nextFormData.file) });
```
Comment (cehsu) [inline src/js/vic-v2/components/PhotoField.jsx]: Is there an advantage of using a this over the file itself as the preview source?

**Review by rianfowler — APPROVED** [review]
LGTM.

File: src/js/vic-v2/components/PhotoField.jsx (hunk 1 above)
Comment (jbalboni) [reply src/js/vic-v2/components/PhotoField.jsx]: As far as I could find, this is what you need to do to use the file itself as the preview source.

**Comment by jbalboni** [comment]
@cehsu @rianfowler This could use another look.

File: src/js/vic-v2/components/PhotoField.jsx [hunk 2]
```diff
@@ -579,12 +612,29 @@ export default class PhotoField extends React.Component {
...
+    this.setState({ previewSrc: src });
+  }
+
   render() {
     const { formData, formContext } = this.props;
     const file = formData || {};
     const onReview = formContext.reviewMode;
 
-    if (onReview) return <PhotoReviewDescription url={getImageUrl(file)}/>;
+    if (onReview) {
```
Comment (rianfowler) [inline src/js/vic-v2/components/PhotoField.jsx]: What is `reviewMode`?

File: src/js/common/schemaform/helpers.js [hunk 3]
```diff
@@ -254,8 +254,8 @@ export function stringifyFormReplacer(key, value) {
     }
 
     // Exclude file data
-    if (value.confirmationCode && value.serverPath) {
-      return _.omit(['serverPath', 'serverName'], value);
+    if (value.confirmationCode && (value.serverPath || value.file)) {
```
Comment (rianfowler) [inline src/js/common/schemaform/helpers.js]: What does `stringifyFormReplacer` do?

File: src/js/common/schemaform/review/ArrayField.jsx [hunk 4]
```diff
@@ -215,6 +216,7 @@ class ArrayField extends React.Component {
                         title={pageTitle}
                         hideTitle
                         name={fieldName}
+                        formContext={formContext}
```
Comment (rianfowler) [inline src/js/common/schemaform/review/ArrayField.jsx]: Can you summarize how this is being passed down now i.e. the hierarchy of components that it's going through? I can see it getting passed through a few schemaform components but I don't see a new utilization of it.

File: src/js/vic-v2/components/PhotoPreview.jsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  render() {
+    const { src, id, className, isLoggedIn, processing } = this.props;
+
+    if (processing) {
+      return (
+        <div className="usa-alert usa-alert-warning vic-processing-warning">
+          <div className="usa-alert-body">
+            <h4 className="usa-alert-title">We're still processing your preview photo</h4>
```
Comment (rianfowler) [inline src/js/vic-v2/components/PhotoPreview.jsx]: My understanding is that we send the photo to our service, that service processes it, and we download the image and use that as the preview.

File: src/js/common/schemaform/helpers.js (hunk 3 above)
Comment (jbalboni) [reply src/js/common/schemaform/helpers.js]: This is used by `transformForSubmit` when stringifying the form data to be submitted to the backend. It's the second parameter for `JSON.stringify`, which lets you modify the data as it's being stringified.

**Review by rianfowler — APPROVED** [review]
LGTM- added some questions for my own understanding

File: src/js/common/schemaform/review/ArrayField.jsx (hunk 4 above)
Comment (jbalboni) [reply src/js/common/schemaform/review/ArrayField.jsx]: `RoutableSavableReviewPage > ReviewPage > ReviewCollapsibleChapter > SchemaForm` and `RoutableSavableReviewPage > ReviewPage > ReviewCollapsibleChapter > ArrayField > SchemaForm` 

Previously we had added to SchemaForm and passed it from `RoutedSavablePage > FormPage > SchemaForm`, so this is to do the wiring for the review page.

File: src/js/vic-v2/components/PhotoField.jsx (hunk 2 above)
Comment (jbalboni) [reply src/js/vic-v2/components/PhotoField.jsx]: On the review page, we render every field in a read only view. Normally there are separate widgets for that (the SchemaForm component choose between them depending on a prop), but for custom fields it's easier to switch between them in the Field component.

File: src/js/vic-v2/components/PhotoPreview.jsx (hunk 5 above)
Comment (jbalboni) [reply src/js/vic-v2/components/PhotoPreview.jsx]: That's what we were doing before, but now the image re-encoding that happens on the backend is done async, so it's possible that we won't be able to get the preview image immediately. Mostly this doesn't matter, because we still have image data in memory, but if you save your form, leave, and come back, it's possible that we'll still be processing.
